### PR TITLE
Path generation

### LINKED ISSUE #610: Path finding

### LINKED ISSUE #611: Hybrid model

### LINKED ISSUE #618: Add com3Hybrid documentation

### PR DESCRIPTION
**Pull request by matthiasto**
- Refactor path computation functions
- Allow to compute path from particles or fields (if from fields needs the FM=FlowMass)
- Runscript to compute a path from com1DFA results (requires that one saves some time steps)
- update hybrid and energy test
- fix #610 
- fix #611
- fix #618

com3HybProfPlot
com3HybRasterPlot

### CODE REVIEW

File: avaframe/com1DFA/com1DFA.py [hunk 1]
```diff
@@ -1093,6 +1085,7 @@ def initializeFields(cfg, dem, particles):
     fields['FV'] = PFV
     fields['P'] = PP
     fields['FD'] = FD
+    fields['FM'] = FD
```
Comment (awirb) [inline avaframe/com1DFA/com1DFA.py:1088]: general comment: do we need to initialize PFV, PP, FD, TA as 2D array of 0s they are all the same here with same dimension, so isn't one enough?

File: avaframe/com1DFA/com1DFACfg.ini [hunk 2]
```diff
@@ -11,7 +11,7 @@ simTypeList = available
 modelType = dfa
 
 #+++++++++++++ Output++++++++++++
-# desired result Parameters (ppr, pfd, pfv, pta, FD, FV, P, Vx, Vy, Vz, TA, particles) - separated by |
+# desired result Parameters (ppr, pfd, pfv, pta, FD, FV, P, FM, Vx, Vy, Vz, TA, particles) - separated by |
```
Comment (awirb) [inline avaframe/com1DFA/com1DFACfg.ini:14]: do we need to add this in the docu somewhere?

File: avaframe/com3Hybrid/hybridModel_com1DFACfg.ini [hunk 3]
```diff
@@ -27,10 +27,13 @@ entThFromShp = False
 entTh = 0.3
```
Comment (awirb) [inline avaframe/com3Hybrid/hybridModel_com1DFACfg.ini:27]: if this is set here consider setting enThFromShp=False?

File: avaframe/in2Trans/shpConversion.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 import copy
 import numpy as np
 import logging
+import os
```
Comment (awirb) [inline avaframe/in2Trans/shpConversion.py:11]: import needed?

File: avaframe/runScripts/runComputeDFAPath.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+"""
+    Create an avalanch path from DFA simulation results
+    Configuration should be specified in DFAPathGenerationCfg.ini
+    or the local version of it.
+    It is possible to generate a path from particles or from fields.
+    From particles, you need to save the particles dictionaries at
+    multiple time steps (first, some in the middle and last)
+    From fields, you need the FD, FM
```
Comment (awirb) [inline avaframe/runScripts/runComputeDFAPath.py:8]: also for multiple time steps

File: avaframe/runScripts/runComputeDFAPath.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+# Start logging
+log = logUtils.initiateLogger(avalancheDir, logName)
+log.info('MAIN SCRIPT')
+log.info('Current avalanche: %s', avalancheDir)
+
+# call com1DFA and perform simulations
+demOri = gI.readDEM(avalancheDir)
+dem = com1DFA.setDEMoriginToZero(demOri)
+dem['originalHeader'] = demOri['header'].copy()
+# Run dense flow model
```
Comment (awirb) [inline avaframe/runScripts/runComputeDFAPath.py:37]: so it is not running com1DFA but just using its results..

File: avaframe/runScripts/runComputeDFAPath.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+# here is an example with com1DFA but another DFA computational module can be used
+# as long as it produces some FV, FD and FM results
+getDFAPathCfgFile = pathlib.Path('ana5Utils', 'DFAPathGenerationCfg.ini')
+getDFAPathCfg = cfgUtils.getModuleConfig(DFAPath)
+
+# create dataFrame of results
+simDF, _ = cfgUtils.readAllConfigurationInfo(avalancheDir)
```
Comment (awirb) [inline avaframe/runScripts/runComputeDFAPath.py:44]: consider already here making a loop over all sims - so it works if you have multiple and if you only have one, which is does now too but just takes the first

**Review by awirb — CHANGES_REQUESTED** [review]

File: avaframe/ana5Utils/DFAPathGeneration.py [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+
+def extendProfileTop(extTopOption, particlesIni, profile):
+    """ extend the DFA path at the top (release)
+
+    Either towards the highest point in particlesIni (extTopOption = 0)
+    or the point leading to the longest runout (extTopOption = 1)
+
+    Parameters
+    -----------
+    extTopOption: boolean
```
Comment (fso42) [inline avaframe/ana5Utils/DFAPathGeneration.py:331]: In the ini extTopOption is not boolean but an integer, with potentially more options, right?

And then I suggest catching if `extTopOption` is something else than 0 or 1 in an else (with an error)

File: avaframe/ana5Utils/DFAPathGeneration.py [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+    profile['x'] = np.append(profile['x'], xExtBottom)
+    profile['y'] = np.append(profile['y'], yExtBottom)
+    profile['z'] = np.append(profile['z'], zExtBottom)
+
+    if debugPlot:
+        debPlot.plotPathExtBot(profile, xInterest, yInterest, 0*yInterest, xLast, yLast)
+    return profile
+
+
+def weightedAvgAndStd(values, weights):
```
Comment (fso42) [inline avaframe/ana5Utils/DFAPathGeneration.py:451]: Might this be an interesting general function in in3Utils?

File: avaframe/com1DFA/com1DFACfg.ini (hunk 2 above)
Comment (fso42) [reply avaframe/com1DFA/com1DFACfg.ini:14]: related to #617

File: avaframe/runScripts/runComputeDFAPath.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    log.info('Computing avalanche path from simulation: %s', simName)
+    pathFromPart = getDFAPathCfg['PATH'].getboolean('pathFromPart')
+    avaProfileMass = DFAPath.generateAveragePath(avalancheDir, pathFromPart, simName, dem)
+    # here the avaProfileMass given in input is overwriten and returns only a x, y, z extended profile
+    avaProfileMass = DFAPath.extendDFAPath(avalancheDir, getDFAPathCfg['PATH'], dem, simName, avaProfileMass)
+    # put path back in original location
+    avaProfileMass['x'] = avaProfileMass['x'] + dem['originalHeader']['xllcenter']
+    avaProfileMass['y'] = avaProfileMass['y'] + dem['originalHeader']['yllcenter']
+    # save profile in Inputs
+    pathAB = pathlib.Path(avalancheDir, 'Outputs', 'DFAPath', 'path_%s_AB_aimec' % simName)
```
Comment (fso42) [inline avaframe/runScripts/runComputeDFAPath.py:54]: Would it make sense to somehow indicate that it is an automatic path by adding something to the name? Like 'massAvgPath_...', 'autoPath...', 'generatedPath'....

**Review by fso42 — CHANGES_REQUESTED** [review]

File: avaframe/runScripts/runComputeDFAPath.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+dem = com1DFA.setDEMoriginToZero(demOri)
+dem['originalHeader'] = demOri['header'].copy()
+# here is an example with com1DFA but another DFA computational module can be used
+# as long as it produces some FV, FD and FM results
+getDFAPathCfgFile = pathlib.Path('ana5Utils', 'DFAPathGenerationCfg.ini')
+getDFAPathCfg = cfgUtils.getModuleConfig(DFAPath)
+
+# create dataFrame of results (read DFA data)
+simDF, _ = cfgUtils.readAllConfigurationInfo(avalancheDir)
+for simName, simDFrow in simDF.iterrows():
```
Comment (fso42) [inline avaframe/runScripts/runComputeDFAPath.py:44]: Catch error when no path/simulations exist in the given dir

**Review by fso42 — CHANGES_REQUESTED** [review]

File: avaframe/runScripts/runComputeDFAPath.py (hunk 11 above)
Comment (fso42) [reply avaframe/runScripts/runComputeDFAPath.py:44]: In general: this needs documentation, maybe right away together with #618

File: avaframe/runScripts/runComputeDFAPath.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+log.info('Current avalanche: %s', avalancheDir)
+
+# call com1DFA and perform simulations
+demOri = gI.readDEM(avalancheDir)
+dem = com1DFA.setDEMoriginToZero(demOri)
+dem['originalHeader'] = demOri['header'].copy()
+# here is an example with com1DFA but another DFA computational module can be used
+# as long as it produces some FV, FD and FM results
+getDFAPathCfgFile = pathlib.Path('ana5Utils', 'DFAPathGenerationCfg.ini')
+getDFAPathCfg = cfgUtils.getModuleConfig(DFAPath)
```
Comment (fso42) [inline avaframe/runScripts/runComputeDFAPath.py:40]: rename this to DFAPathCfg: the get is a bit confusing here...

**Review by fso42 — CHANGES_REQUESTED** [review]

**Review by fso42 — CHANGES_REQUESTED** [review]

**Review by fso42 — CHANGES_REQUESTED** [review]

**Review by fso42 — APPROVED** [review]

**Pull request merged**